### PR TITLE
[runtime] Implement support for SuppressGCTransitionAttribute.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/mono/mono/issues/17367.

### CODE REVIEW

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by filipnavara** [comment]
Is it somehow guaranteed that GC poll is inserted after the P/Invoke calls with this attribute?

**Comment by lambdageek** [comment]
> Is it somehow guaranteed that GC poll is inserted after the P/Invoke calls with this attribute?

I don't think so.  I'm not sure it's necessary.  We will get the usual safepoints based on the structure of the caller (ie, loops, exn handlers, fn entry)

**Comment by filipnavara** [comment]
@lambdageek It was a concern in the [original PR](https://github.com/dotnet/coreclr/pull/26458) that introduced the attribute to CoreCLR. I believe that GC polling at loops should be good enough but I may have missed something. The CoreCLR code definitely inserts a GC poll after the P/Invoke call marked with the attribute.

**Comment by lambdageek** [comment]
Thanks!

The discussion starting from https://github.com/dotnet/coreclr/pull/26458#issuecomment-527260072 and especially this part by jkotas two comments later:

> The problem is composition. This will be used to implement library APIs and same restrictions would transitively apply to the library APIs that use this as an implementation detail. For example, if this is used for QueryPerformanceCounter(), the Stopwatch documentation would need to include a warning that calling Stopwatch too often will lead to long GC pause times. That would be super confusing.

@vargaz I think we need a safepoint, too.  If we add one after the icall, will LLVM avoid adding one at loop entry?

**Comment by vargaz** [comment]
There is a already a safepoint in the prolog of the wrapper, like in all methods.

File: mono/metadata/marshal.c [hunk 1]
```diff
@@ -111,6 +111,10 @@ delegate_hash_table_remove (MonoDelegate *d);
 GENERATE_TRY_GET_CLASS_WITH_CACHE (stringbuilder, "System.Text", "StringBuilder");
 static GENERATE_TRY_GET_CLASS_WITH_CACHE (unmanaged_function_pointer_attribute, "System.Runtime.InteropServices", "UnmanagedFunctionPointerAttribute");
 
+#ifdef ENABLE_NETCORE
```
Comment (jaykrell) [inline mono/metadata/marshal.c]: Isn't it trivial to expose this in non-netcore also, w/o breaking anything?
Just remove the `#if`s?

**Pull request merged**